Zig and Node.js N-API example

Given Zig's promise of C header includes, it would be cool for the docs or blog to have an example of writing a "Hello Node.js" native addon for Node.js using N-API and Zig.

Node's N-API is fairly recent, and many existing native addons for Node written using Nan etc. are probably being ported. Now might be a great time for some of these native Node modules to consider using Zig instead of C/C++.

Hi @ifreund  - thanks for the example! I ran into some problems trying to get it to run zig 0.9.0-dev.1583+a7d215759:

- it seems like the compiler complains about `info` being unused in `foo`, so I changed it to `_`
- the compiler complains about `.napi_ok` so I just replaced them with 0 (I first tried `@enumToInt(...)` but that didn't seem to work)
```
./example.zig:7:68: error: incompatible types: 'c_uint' and '(enum literal)'
    if (c.napi_create_function(env, null, 0, foo, null, &function) != .napi_ok) {
 ```

I then ended up with the following code:

```zig
const c = @cImport({
    @cInclude("node_api.h");
});

export fn napi_register_module_v1(env: c.napi_env, exports: c.napi_value) c.napi_value {
    var function: c.napi_value = undefined;
    if (c.napi_create_function(env, null, 0, foo, null, &function) != 0) {
        _ = c.napi_throw_error(env, null, "Failed to create function");
        return null;
    }

    if (c.napi_set_named_property(env, exports, "foo", function) != 0) {
        _ = c.napi_throw_error(env, null, "Failed to add function to exports");
        return null;
    }

    return exports;
}

fn foo(env: c.napi_env, _: c.napi_callback_info) callconv(.C) c.napi_value {
    var result: c.napi_value = undefined;
    if (c.napi_create_int32(env, 42, &result) != 0) {
        _ = c.napi_throw_error(env, null, "Failed to create return value");
        return null;
    }

    return result;
}
```

I tried running this on Debian Linux and it seems to work as expected (nice!):

```
$ zig build-lib -dynamic -lc -isystem $INCLUDE example.zig -femit-bin=example.node
$ node example.js
42
```

However, I was not successful in getting this to run on macOS 12.0.1:

```
$ zig build-lib -dynamic -lc -isystem $INCLUDE example.zig -femit-bin=example.node
error(link): undefined reference to symbol '_napi_throw_error'
error(link):   first referenced in '/tmp/zig-cache/o/910087189ff92cf8ebf244f7e7a8e01b/example.o'
error(link): undefined reference to symbol '_napi_set_named_property'
error(link):   first referenced in '/tmp/zig-cache/o/910087189ff92cf8ebf244f7e7a8e01b/example.o'
error(link): undefined reference to symbol '_napi_create_int32'
error(link):   first referenced in '/tmp/zig-cache/o/910087189ff92cf8ebf244f7e7a8e01b/example.o'
error(link): undefined reference to symbol '_napi_create_function'
error(link):   first referenced in '/tmp/zig-cache/o/910087189ff92cf8ebf244f7e7a8e01b/example.o'
error: UndefinedSymbolReference
```

https://github.com/kristoff-it/zig-cuckoofilter/blob/master/c-abi-examples/node_example.js#L13-L17 seems to imply macOS might need a different approach, but I tried doing a `build-obj` and running `gcc` as described there and couldn't get anything workable. Is this a known issue? Is it possible you could update your example with the latest and greatest commands for macOS as well? Thanks!

> $ zig build-lib -dynamic -lc -isystem $INCLUDE example.zig -femit-bin=example.node
> error(link): undefined reference to symbol '_napi_throw_error'
> error(link):   first referenced in '/tmp/zig-cache/o/910087189ff92cf8ebf244f7e7a8e01b/example.o'
> error(link): undefined reference to symbol '_napi_set_named_property'
> error(link):   first referenced in '/tmp/zig-cache/o/910087189ff92cf8ebf244f7e7a8e01b/example.o'
> error(link): undefined reference to symbol '_napi_create_int32'
> error(link):   first referenced in '/tmp/zig-cache/o/910087189ff92cf8ebf244f7e7a8e01b/example.o'
> error(link): undefined reference to symbol '_napi_create_function'
> error(link):   first referenced in '/tmp/zig-cache/o/910087189ff92cf8ebf244f7e7a8e01b/example.o'
> error: UndefinedSymbolReference

This is the same error you'd traditionally get for C if you forgot to specify `--allow-shlib-undefined` for linux; or more specifically for OSX: `-undefined dynamic_lookup`.
I think this is covered by https://github.com/ziglang/zig/issues/8180

--- a/src/link/MachO.zig
+++ b/src/link/MachO.zig
@@ -444,6 +444,7 @@ pub fn flushModule(self: *MachO, comp: *Compilation) !void {
     const is_dyn_lib = self.base.options.link_mode == .Dynamic and is_lib;
     const is_exe_or_dyn_lib = is_dyn_lib or self.base.options.output_mode == .Exe;
     const stack_size = self.base.options.stack_size_override orelse 0;
+    const allow_undef = is_dyn_lib and (self.base.options.allow_shlib_undefined orelse false);
 
     const id_symlink_basename = "zld.id";
 
@@ -847,6 +848,11 @@ pub fn flushModule(self: *MachO, comp: *Compilation) !void {
                     try argv.append(try std.fmt.allocPrint(arena, "-F{s}", .{framework_dir}));
                 }
 
+                if (allow_undef) {
+                    try argv.append("-undefined");
+                    try argv.append("dynamic_lookup");
+                }
+
                 Compilation.dump_argv(argv.items);
             }
 
@@ -897,6 +903,16 @@ pub fn flushModule(self: *MachO, comp: *Compilation) !void {
                         .n_desc = 0,
                         .n_value = 0,
                     };
+                    _ = self.unresolved.swapRemove(resolv.where_index);
+                    continue;
+                } else if (allow_undef) {
+                    const n_desc = @bitCast(
+                        u16,
+                        macho.BIND_SPECIAL_DYLIB_FLAT_LOOKUP * @intCast(i16, macho.N_SYMBOL_RESOLVER),
+                    );
+                    // TODO allow_shlib_undefined is an ELF flag so figure out macOS specific flags too.
+                    sym.n_type = macho.N_EXT;
+                    sym.n_desc = n_desc;
                     _ = self.unresolved.swapRemove(resolv.where_index);
                     continue;
                 }
@@ -5111,7 +5127,7 @@ fn writeDyldInfoData(self: *MachO) !void {
                             try bind_pointers.append(.{
                                 .offset = binding.offset + base_offset,
                                 .segment_id = match.seg,
-                                .dylib_ordinal = @divExact(bind_sym.n_desc, macho.N_SYMBOL_RESOLVER),
+                                .dylib_ordinal = @divExact(@bitCast(i16, bind_sym.n_desc), macho.N_SYMBOL_RESOLVER),
                                 .name = self.getString(bind_sym.n_strx),
                             });
                         },
@@ -5133,7 +5149,7 @@ fn writeDyldInfoData(self: *MachO) !void {
                             try lazy_bind_pointers.append(.{
                                 .offset = binding.offset + base_offset,
                                 .segment_id = match.seg,
-                                .dylib_ordinal = @divExact(bind_sym.n_desc, macho.N_SYMBOL_RESOLVER),
+                                .dylib_ordinal = @divExact(@bitCast(i16, bind_sym.n_desc), macho.N_SYMBOL_RESOLVER),
                                 .name = self.getString(bind_sym.n_strx),
                             });
                         },

--- a/src/main.zig
+++ b/src/main.zig
@@ -1703,6 +1703,16 @@ fn buildOutputType(
                     }
                     emit_implib = .{ .yes = linker_args.items[i] };
                     emit_implib_arg_provided = true;
+                } else if (mem.eql(u8, arg, "-undefined")) {
+                    i += 1;
+                    if (i >= linker_args.items.len) {
+                        fatal("expected linker arg after '{s}'", .{arg});
+                    }
+                    if (mem.eql(u8, "dynamic_lookup", linker_args.items[i])) {
+                        linker_allow_shlib_undefined = true;
+                    } else {
+                        fatal("unsupported -undefined option '{s}'", .{linker_args.items[i]});
+                    }
                 } else {
                     warn("unsupported linker arg: {s}", .{arg});
                 }

